sys/net/network_layer/ng_sixlowpan/frag/ng_sixlowpan_frag.c

```diff
@@ -0,0 +1,233 @@
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+
+#include "kernel_types.h"
+#include "net/ng_pktbuf.h"
+#include "net/ng_netapi.h"
+#include "net/ng_netif/hdr.h"
+#include "net/ng_sixlowpan/frag.h"
+#include "net/ng_sixlowpan/netif.h"
+#include "utlist.h"
+
+#include "rbuf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static uint16_t _tag = 0;
+
+static inline uint16_t _floor8(uint16_t length)
+{
+    return length & 0xf8U;
+}
+
+static inline size_t _min(size_t a, size_t b)
+{
+    return (a < b) ? a : b;
+}
+
+static ng_pktsnip_t *_build_frag_pkt(ng_pktsnip_t *pkt, size_t payload_len,
+                                     size_t size)
+{
+    ng_netif_hdr_t *hdr = pkt->data, *new_hdr;
+    ng_pktsnip_t *netif, *frag;
+
+    netif = ng_netif_hdr_build(ng_netif_hdr_get_src_addr(hdr), hdr->src_l2addr_len,
+                               ng_netif_hdr_get_dst_addr(hdr), hdr->dst_l2addr_len);
+
+    if (netif == NULL) {
+        DEBUG("6lo frag: error allocating new link-layer header\n");
+        return NULL;
+    }
+
+    new_hdr = netif->data;
+    new_hdr->if_pid = hdr->if_pid;
+    new_hdr->flags = hdr->flags;
+    new_hdr->rssi = hdr->rssi;
+    new_hdr->lqi = hdr->lqi;
+
+    frag = ng_pktbuf_add(NULL, NULL, _min(size, payload_len),
+                         NG_NETTYPE_SIXLOWPAN);
+
+    if (frag == NULL) {
+        DEBUG("6lo frag: error allocating first fragment\n");
+        ng_pktbuf_release(netif);
+        return NULL;
+    }
+
+    LL_PREPEND(frag, netif);
+
+    return frag;
+}
+
+static uint16_t _send_1st_fragment(ng_sixlowpan_netif_t *iface, ng_pktsnip_t *pkt,
+                                   size_t payload_len, size_t datagram_size)
+{
+    ng_pktsnip_t *frag;
+    uint16_t max_frag_size = _floor8(iface->max_frag_size -
+                                     (payload_len - datagram_size) -
+                                     sizeof(ng_sixlowpan_frag_t));
+    uint16_t local_offset = 0;
+    ng_sixlowpan_frag_t *hdr;
+    uint8_t *data;
+
+    DEBUG("6lo frag: determined max_frag_size = %" PRIu16 "\n", max_frag_size);
+
+    /* 6LoWPAN dispatches don't count into that */
+    max_frag_size += (payload_len - datagram_size);
+
+    frag = _build_frag_pkt(pkt, payload_len,
+                           max_frag_size + sizeof(ng_sixlowpan_frag_t));
+
+    if (frag == NULL) {
+        return 0;
+    }
+
+    hdr = frag->next->data;
+    data = (uint8_t *)(hdr + 1);
+
+    hdr->disp_size = byteorder_htons((uint16_t)datagram_size);
+    hdr->disp_size.u8[0] |= NG_SIXLOWPAN_FRAG_1_DISP;
+    hdr->tag = byteorder_htons(_tag);
+
+    pkt = pkt->next;    /* don't copy netif header */
+
+    while (pkt != NULL) {
+        size_t clen = _min(max_frag_size - local_offset, pkt->size);
+
+        memcpy(data + local_offset, pkt->data, clen);
+        local_offset += clen;
+
+        if (local_offset >= max_frag_size) {
+            break;
+        }
+
+        pkt = pkt->next;
```

Comment:
I had an explanation for that, but forgot it… let me check that. Otherwise you might be right.